### PR TITLE
Audio:  ASRC: Code reorganization

### PR DESCRIPTION
**Pull request by andrula-song**
1. Move all ASRC related header files into src/audio/asrc folder.
2. split IPC version related code to asrc_ipc3.c and asrc_ipc4.c.

### CODE REVIEW

File: src/audio/asrc/asrc_ipc3.c [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <rtos/string.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "asrc.h"
```
Comment (btian1) [inline src/audio/asrc/asrc_ipc3.c:29]: please remove unused headers.  see tdfb_ipc3.c

**Comment by keqiaozhang** [comment]
SOFCI TEST

File: src/audio/CMakeLists.txt [hunk 2]
```diff
@@ -174,19 +174,20 @@ set(sof_audio_modules mixer volume src asrc eq-fir eq-iir dcblock crossover tdfb
...
 if(CONFIG_IPC_MAJOR_3)
 	set(volume_sources volume/volume.c volume/volume_generic.c volume/volume_ipc3.c)
+	set(asrc_sources asrc/asrc.c asrc/asrc_farrow.c asrc/asrc_farrow_generic.c asrc/asrc_ipc3.c)
 	set(src_sources src/src.c src/src_ipc3.c src/src_generic.c)
 	set(eq-iir_sources eq_iir/eq_iir_ipc3.c eq_iir/eq_iir_generic.c)
 	set(eq-fir_sources eq_fir/eq_fir_ipc3.c)
 	set(tdfb_sources tdfb/tdfb_ipc3.c)
 elseif(CONFIG_IPC_MAJOR_4)
 	set(volume_sources volume/volume.c volume/volume_generic.c volume/volume_ipc4.c)
+	set(asrc_sources asrc/asrc.c asrc/asrc_farrow.c asrc/asrc_farrow_generic.c asrc/asrc_ipc4.c)
```
Comment (RanderWang) [inline src/audio/CMakeLists.txt:184]: please check eq-fir_sources for example:  share the common "set(asrc_sources asrc/asrc.c asrc/asrc_farrow.c asrc/asrc_farrow_generic.c)" after this version check , only set "asrc/asrc_ipc3.c or asrc/asrc_ipc4.c" here,

**Review by RanderWang — COMMENTED** [review]
LGTM

File: src/audio/asrc/asrc.c [hunk 3]
```diff
@@ -268,94 +201,7 @@ static void src_copy_s16(struct processing_module *mod,
...
-}
-
-static inline bool asrc_get_asynchronous_mode(const struct ipc4_asrc_module_cfg *ipc_asrc)
-{
-	return false;
-}
-
-#endif
-
-static int asrc_init(struct processing_module *mod)
```
Comment (btian1) [inline src/audio/asrc/asrc.c:318]: asrc_init is standard interface for module interface, need keep in asrc.c, only move out specific code with like asrc_common_init to cover both ipc3/ipc4 cases.

File: src/audio/asrc/asrc.c [hunk 4]
```diff
@@ -465,17 +311,15 @@ static int asrc_verify_params(struct processing_module *mod,
...
+				 params->rate, cd->source_rate);
 			return -EINVAL;
 		}
 	} else {
-		if (asrc_get_sink_rate(&cd->ipc_config) &&
-		    params->rate != asrc_get_sink_rate(&cd->ipc_config)) {
+		if (cd->sink_rate && params->rate != cd->sink_rate) {
 			comp_err(dev, "asrc_verify_params(): runtime stream pcm rate %u does not match rate %u fetched from ipc.",
-				 params->rate, asrc_get_sink_rate(&cd->ipc_config));
+				 params->rate, cd->sink_rate);
```
Comment (btian1) [inline src/audio/asrc/asrc.c:322]: no need to change line 314 - 322, it is not part of code clean or re-organize.
if you want to change, you can use a separate pr with full investigation.

File: src/audio/asrc/asrc.c [hunk 5]
```diff
@@ -502,9 +346,7 @@ static int asrc_params(struct processing_module *mod)
 
 	comp_info(dev, "asrc_params()");
 
-#if CONFIG_IPC_MAJOR_4
-	ipc4_base_module_cfg_to_stream_params(&cd->ipc_config.base, pcm_params);
-#endif
+	set_stream_params(cd, pcm_params);
```
Comment (btian1) [inline src/audio/asrc/asrc.c:349]: need change function name to:asrc_set_common_stream_params

File: src/audio/asrc/asrc.c [hunk 6]
```diff
@@ -517,15 +359,13 @@ static int asrc_params(struct processing_module *mod)
 	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
 				source_list);
 
-#if CONFIG_IPC_MAJOR_4
 	/* update the source/sink buffer formats. Sink rate will be modified below */
-	ipc4_update_buffer_format(sourceb, &cd->ipc_config.base.audio_fmt);
-	ipc4_update_buffer_format(sinkb, &cd->ipc_config.base.audio_fmt);
-#endif
+	update_buffer_format(sourceb, cd);
+	update_buffer_format(sinkb, cd);
```
Comment (btian1) [inline src/audio/asrc/asrc.c:364]: add asrc_ prefix for function name.

File: src/audio/asrc/asrc.c [hunk 7]
```diff
@@ -517,15 +359,13 @@ static int asrc_params(struct processing_module *mod)
...
 
 	/* Don't change sink rate if value from IPC is 0 (auto detect) */
-	if (asrc_get_sink_rate(&cd->ipc_config))
-		audio_stream_set_rate(&sinkb->stream, asrc_get_sink_rate(&cd->ipc_config));
+	if (cd->sink_rate)
+		audio_stream_set_rate(&sinkb->stream, cd->sink_rate);
```
Comment (btian1) [inline src/audio/asrc/asrc.c:368]: no need change, also PR title is not code clean, it is code reorganize.

File: src/audio/asrc/asrc.h [hunk 8]
```diff
@@ -165,4 +169,73 @@ struct ipc4_asrc_module_cfg {
...
+int asrc_dai_start_timestamp(struct comp_data *cd);
+int asrc_dai_stop_timestamp(struct comp_data *cd);
+#if CONFIG_IPC_MAJOR_4 && CONFIG_ZEPHYR_NATIVE_DRIVERS
+int asrc_dai_get_timestamp(struct comp_data *cd, struct dai_ts_data *tsd);
+#else
+int asrc_dai_get_timestamp(struct comp_data *cd, struct timestamp_data *tsd);
+#endif
+void update_buffer_format(struct comp_buffer *buf_c, struct comp_data *cd);
+void set_stream_params(struct comp_data *cd, struct sof_ipc_stream_params *params);
+extern const struct sof_uuid asrc_uuid;
```
Comment (btian1) [inline src/audio/asrc/asrc.h:238]: I did not add those extern, should no need for this.

File: src/audio/asrc/asrc.c [hunk 9]
```diff
@@ -268,94 +201,7 @@ static void src_copy_s16(struct processing_module *mod,
...
-
-static int asrc_init(struct processing_module *mod)
-{
-	struct comp_dev *dev = mod->dev;
-	struct module_data *mod_data = &mod->priv;
-#ifndef CONFIG_IPC_MAJOR_4
-	const struct ipc_config_asrc *ipc_asrc = mod_data->cfg.init_data;
-#else
-	const struct ipc4_asrc_module_cfg *ipc_asrc =  mod_data->cfg.init_data;
-#endif
```
Comment (lyakh) [inline src/audio/asrc/asrc.c:326]: seems like an overkill to duplicate the whole function just because of one type mismatch. Maybe use a `typedef` or even just a `#define` in a header, e.g.
#ifdef CONFIG_IPC_MAJOR_4
typedef struct ipc4_asrc_module_cfg ipc_asrc_cfg;
#else
typedef struct ipc_config_asrc ipc_asrc_cfg;
#endif
```

File: src/audio/asrc/asrc.c (hunk 4 above)
Comment (andrula-song) [reply src/audio/asrc/asrc.c:322]: as I told you before, actually we will can use the parameters form init to replace the function call here.

File: src/audio/asrc/asrc.h (hunk 8 above)
Comment (andrula-song) [reply src/audio/asrc/asrc.h:238]: look at https://github.com/thesofproject/sof/pull/8161

**Comment by andrula-song** [comment]
SOFCI TEST

File: src/audio/asrc/asrc.h (hunk 8 above)
Comment (btian1) [reply src/audio/asrc/asrc.h:281]: yes, src/asrc have different uuid for ipc3 and ipc4, seems this is the only way.

File: src/audio/asrc/asrc.c (hunk 7 above)
Comment (lgirdwood) [reply src/audio/asrc/asrc.c:410]: yeah - nothing wrong with the cleanup, we should just put into into the next patch so we can bisect if an issue is introduced.

**Review by lgirdwood — COMMENTED** [review]
LGTM, needs the cleanups in a different patch.

File: src/audio/asrc/asrc_ipc4.c [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+/* 66b4402d-b468-42f2-81a7-b37121863dd4 */
+DECLARE_SOF_RT_UUID("asrc", asrc_uuid, 0x66b4402d, 0xb468, 0x42f2,
+		    0x81, 0xa7, 0xb3, 0x71, 0x21, 0x86, 0x3d, 0xd4);
+
+DECLARE_TR_CTX(asrc_tr, SOF_UUID(asrc_uuid), LOG_LEVEL_INFO);
+
+int asrc_dai_configure_timestamp(struct comp_data *cd)
+{
+	if (cd->dai_dev) {
```
Comment (lyakh) [inline src/audio/asrc/asrc_ipc4.c:27]: I think all these would look simpler with
```
if (!cd->dai_dev)
    return -ENODEV;
```
(or whatever error code you prefer)

**Review by lyakh — COMMENTED** [review]
I like it much better with the `typedef`!

File: src/audio/CMakeLists.txt [hunk 11]
```diff
@@ -174,12 +174,14 @@ set(sof_audio_modules mixer volume src asrc eq-fir eq-iir dcblock crossover tdfb
 # sources for each module
 if(CONFIG_IPC_MAJOR_3)
 	set(volume_sources volume/volume.c volume/volume_generic.c volume/volume_ipc3.c)
+	set(asrc_sources asrc/asrc_ipc3.c)
```
[the remark this answers is not in the document]
Comment (btian1) [reply src/audio/CMakeLists.txt:177]: @lgirdwood , this change is based on search, so I request use a separate PR to track clean up src/audio/cMakeList.txt

File: zephyr/CMakeLists.txt [hunk 12]
```diff
@@ -482,13 +482,23 @@ zephyr_library_sources_ifdef(CONFIG_MATH_IIR_DF2T
...
+		${SOF_AUDIO_PATH}/asrc/asrc_farrow.c
+		${SOF_AUDIO_PATH}/asrc/asrc_farrow_generic.c
+	)
+elseif(CONFIG_IPC_MAJOR_4)
+	zephyr_library_sources_ifdef(CONFIG_COMP_ASRC
+		${SOF_AUDIO_PATH}/asrc/asrc.c
+		${SOF_AUDIO_PATH}/asrc/asrc_ipc4.c
+		${SOF_AUDIO_PATH}/asrc/asrc_farrow_hifi3.c
+		${SOF_AUDIO_PATH}/asrc/asrc_farrow.c
+		${SOF_AUDIO_PATH}/asrc/asrc_farrow_generic.c
```
Comment (marc-hb) [inline zephyr/CMakeLists.txt:499]: Copy/paste, seriously? Move the common .c files out of the CONFIG_IPC_.

Don't nest `CONFIG_COMP_ASRC` inside `CONFIG_IPC_MAJOR_x`, do it the other way round.

**Review by marc-hb — CHANGES_REQUESTED** [review]

File: src/audio/asrc/asrc_ipc4.c (hunk 10 above)
Comment (marc-hb) [reply src/audio/asrc/asrc_ipc4.c]: It's called "Early Return".

https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

It reduces indentation and is much easier to read. It's similar to `assert()`.

Comment (andrula-song) [reply src/audio/asrc/asrc_ipc4.c]: agree, done.

File: zephyr/CMakeLists.txt (hunk 12 above)
Comment (andrula-song) [reply zephyr/CMakeLists.txt:499]: or should I move it to line 718 like you did in https://github.com/thesofproject/sof/pull/8425?
`if(CONFIG_IPC_MAJOR_3)
	zephyr_library_sources_ifdef(CONFIG_SAMPLE_SMART_AMP
		${SOF_SAMPLES_PATH}/audio/smart_amp_test_ipc3.c
	)

	zephyr_library_sources_ifdef(CONFIG_COMP_TDFB
	${SOF_AUDIO_PATH}/tdfb/tdfb_ipc3.c
	)
elseif(CONFIG_IPC_MAJOR_4)
	zephyr_library_sources_ifdef(CONFIG_SAMPLE_SMART_AMP
		${SOF_SAMPLES_PATH}/audio/smart_amp_test_ipc4.c
	)

	zephyr_library_sources_ifdef(CONFIG_COMP_TDFB
	${SOF_AUDIO_PATH}/tdfb/tdfb_ipc4.c
	)
endif()`
or line 452?
there are several styles in this document.  the volume & src & module_adapter is copy/past, and tdfb & smart_amp_test & eq is another format, and if I nest CONFIG_IPC_MAJOR_x inside CONFIG_COMP_ASRC then it would be third style

File: src/audio/asrc/asrc.h [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+ * The low pass function is y(n) = c1 * x(n) + c2 * y(n -1)
+ * coefficient c2 needs to be 1 - c1.
+ */
+#define COEF_C1		Q_CONVERT_FLOAT(0.01, 30)
+#define COEF_C2		Q_CONVERT_FLOAT(0.99, 30)
+
+#ifdef CONFIG_IPC_MAJOR_4
+typedef struct ipc4_asrc_module_cfg ipc_asrc_cfg;
+#else
+typedef struct ipc_config_asrc ipc_asrc_cfg;
```
Comment (btian1) [inline src/audio/asrc/asrc.h:188]: seems code style check does not allow this, is it must fix @kv2019i ?

File: src/audio/asrc/asrc.c [hunk 14]
```diff
@@ -272,58 +201,11 @@ static void src_copy_s16(struct processing_module *mod,
...
 static int asrc_init(struct processing_module *mod)
 {
 	struct comp_dev *dev = mod->dev;
 	struct module_data *mod_data = &mod->priv;
-#ifndef CONFIG_IPC_MAJOR_4
-	const struct ipc_config_asrc *ipc_asrc = mod_data->cfg.init_data;
-#else
-	const struct ipc4_asrc_module_cfg *ipc_asrc =  mod_data->cfg.init_data;
-#endif
+	const ipc_asrc_cfg *ipc_asrc = mod_data->cfg.init_data;
```
Comment (btian1) [inline src/audio/asrc/asrc.c:208]: const ipc_asrc_cfg *ipc_asrc = (const ipc_asrc_cfg *)mod_data->cfg.init_data;

File: src/audio/asrc/asrc.h [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+	int sink_frames_max;	/* Max # of frames to process at sink */
+	int data_shift;		/* Optional shift by 8 to process S24_4LE */
+	uint8_t *buf;		/* Samples buffer for input and output */
+	uint8_t *ibuf[PLATFORM_MAX_CHANNELS];	/* Input channels pointers */
+	uint8_t *obuf[PLATFORM_MAX_CHANNELS];	/* Output channels pointers */
+	bool track_drift;
+	asrc_proc_func asrc_func;		/* ASRC processing function */
+};
+
+#ifndef CONFIG_IPC_MAJOR_4
```
Comment (btian1) [inline src/audio/asrc/asrc.h:229]: do you think below is a good method?
combine all CONFIG_IPC_MAJOR_4 CONFIG_IPC_MAJOR_3
with only once in header file, and move all related inside.
#if CONFIG_IPC_MAJOR_4

#8034 
#if CONFIG_IPC_MAJOR_3
#endif

any comments?

**Review by btian1 — APPROVED** [review]
ple

**Review by RanderWang — COMMENTED** [review]
Why we need to move it to ASRC folder ? Thanks!

File: src/audio/asrc/asrc.h (hunk 15 above)
Comment (lgirdwood) [reply src/audio/asrc/asrc.h:229]: This is fine for now. We can incrementally update if needed.

File: zephyr/CMakeLists.txt (hunk 12 above)
Comment (marc-hb) [reply zephyr/CMakeLists.txt:499]: It does matter to me a great lot because tracking all these configuration inconsistencies is what makes the cleanup very time-consuming. It would be very quick and easy otherwise.

File: src/audio/asrc/asrc.h [hunk 16]
```diff
@@ -17,10 +17,42 @@
...
+#define ASRC_HIFI3	0
+#endif /* XCC */
+#else
+/* Applied when ASRC_AUTOARCH is set to zero */
+#define ASRC_GENERIC	1 /* Enable generic */
+#define ASRC_HIFI3	0 /* Disable HiFi3  */
+#endif /* Autoarch */
+
+#if CONFIG_IPC_MAJOR_4
+
```
Comment (plbossart) [inline src/audio/asrc/asrc.h:59]: This part is inconsistent, you are splitting the code in IPC4 and IPC3-specific parts, and create a common header which has IPC4-specific stuff.

If you truly want to have IPC-agnostic ASRC, this needs to be updated.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: src/audio/asrc/asrc.h (hunk 16 above)
Comment (andrula-song) [reply src/audio/asrc/asrc.h:59]: thanks, done.

**Comment by btian1** [comment]
@lyakh , with typedef, there are codestyle errors, do you still insist use typedef, how about #define, or directly use # if IPC3?

**Comment by lyakh** [comment]
> @lyakh , with typedef, there are codestyle errors, do you still insist use typedef, how about #define, or directly use # if IPC3?

@btian1 I never insisted on that, it was my proposal. I still think it's ok to ignore code style complaints in this case, but if others disagree we can also use a `#define` which would be my second preference.

**Review by lyakh — APPROVED** [review]

File: src/audio/asrc/asrc_config.h [hunk 17]
```diff
@@ -16,7 +16,7 @@
 #if defined __XCC__
 /* For xt-xcc */
 #include <xtensa/config/core-isa.h>
-#if XCHAL_HAVE_HIFI3 == 1
+#if XCHAL_HAVE_HIFI3 == 1 || XCHAL_HAVE_HIFI4 == 1
```
Comment (plbossart) [inline src/audio/asrc/asrc_config.h:19]: this isn't really a code reorganization... It's really not good practice to add such configuration changes in the middle or a code move...

This should be in a separate patch IMHO.

File: src/audio/asrc/asrc_ipc3.c [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+{
+	if (!cd->dai_dev)
+		return -EINVAL;
+
+	return cd->dai_dev->drv->ops.dai_ts_get(cd->dai_dev, tsd);
+}
+
+void asrc_update_buffer_format(struct comp_buffer *buf_c, struct comp_data *cd)
+{
+	//IPC3 don't need to update audio stream format here.
```
Comment (kv2019i) [inline src/audio/asrc/asrc_ipc3.c:59]: C-style comments would be better

File: src/audio/asrc/asrc_ipc3.c [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+void asrc_set_stream_params(struct comp_data *cd, struct sof_ipc_stream_params *params)
+{
+	//IPC3 don't need to set audio stream params here.
```
Comment (kv2019i) [inline src/audio/asrc/asrc_ipc3.c:64]: Ditto here

**Review by kv2019i — COMMENTED** [review]
Mostly look good, a few minor comments inline.

File: src/audio/CMakeLists.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (andrula-song) [reply src/audio/CMakeLists.txt:180]: actually, it is set on line 196 as `"set(asrc_sources asrc/asrc.c asrc/asrc_farrow.c asrc/asrc_farrow_generic.c)"`.  Adding here is just keep the asrc_sources be right after I refine ASRC code.

**Review by singalsu — COMMENTED** [review]
Looks good, thanks! With assumption Kai's findings fixed.

**Review by kv2019i — APPROVED** [review]
